### PR TITLE
import unenv-preset 

### PR DESCRIPTION
**Pull request by vicb**
@pi0 @petebacondarwin This PR imports & sync the cloudflare preset.

https://github.com/cloudflare/workers-sdk/issues/7366 lists (most of) the tasks that will need to be done.

--

- imported from https://github.com/pi0/unenv-preset-cloudflare
  - sync'd with unjs/unenv at d9d4d03
  - added back the .cjs to fix the build

Smoke tested with https://github.com/cloudflare/workers-sdk/tree/preset-test. In that branche, wrangler uses `@cloudflare/unenv-preset` and the hybrid test pass.

Next steps:
- Add a way to test the dev version of the package (the plugin should resolve modules to the workspace version of the file)
- Add tests
- Setup the NPM release
- ... see tracking issue

--

@pi0 I changed the license to MIT + Apache, is that ok with you?

@petebacondarwin do you think it would cause an issue to merge this before all of #7366 is done?

- Tests
  - [ ] TODO (before merge)
  - [ ] Tests included
  - [x] Tests not necessary because: will be added in a follow up PR
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [ ] Required
  - [x] Not required because: newly created pkg not used yet
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): 
  - [x] Documentation not necessary because: there is no user facing change

### CODE REVIEW

File: packages/unenv-preset/build.config.ts
Comment (pi0) [inline packages/unenv-preset/build.config.ts:1]: while personally recommand unbuild/mkdist, also a simple tsc would work if it is widely used in this repo.

File: packages/unenv-preset/eslint.config.mjs
Comment (pi0) [inline packages/unenv-preset/eslint.config.mjs:1]: Do you have any other convention perhaps here?

**Comment by pi0** [comment]
@vicb I think you need to update [deps.ts](https://github.com/cloudflare/workers-sdk/blob/98d27250d2b37451e808b4aee623c3b4c4de5a20/packages/wrangler/scripts/deps.ts#L21) to include this new package.

**Comment by vicb** [comment]
> @vicb I think you need to update [deps.ts](https://github.com/cloudflare/workers-sdk/blob/98d27250d2b37451e808b4aee623c3b4c4de5a20/packages/wrangler/scripts/deps.ts#L21) to include this new package.

Good point, thanks @pi0. I'll add that to the tracking issue for when we switch wrangler to using this package.

File: .gitignore
Comment (anonrig) [inline .gitignore:201]: Can we add a codeowners document as well?

File: packages/unenv-preset/package.json
Comment (anonrig) [inline packages/unenv-preset/package.json:48]: Should this come from workspace as well?

File: .gitignore
Comment (vicb) [reply .gitignore:201]: @anonrig Add a task in the tracking issue to setup CODEOWNERS

File: packages/unenv-preset/package.json
Comment (vicb) [reply packages/unenv-preset/package.json:48]: Nope.
This is the version of the types matching the unenv impl.

File: packages/unenv-preset/src/index.ts
Comment (pi0) [inline packages/unenv-preset/src/index.ts:2]: Is is likely that this will change before the final unenv v2 release to simpler `createEnv({ presets: [], ...options })` and we drop/deprecate older exports.

Totally up to you (as long as we have unenv as relaxed `^` peer dependency it is fine) but I would very much root for not doing reexport and compose where preset is "consumed" (it can be `wrangler`, esm.sh, nitro or other places using unenv and will be always forward compatible)

File: packages/unenv-preset/src/runtime/node/async_hooks/$cloudflare.ts
Comment (pi0) [inline packages/unenv-preset/src/runtime/node/async_hooks/$cloudflare.ts:1]: You can convert `$cloudflare` files to `index` btw now as it is an actual external preset. IMO it can make things more readable.

File: packages/unenv-preset/src/index.ts
Comment (vicb) [reply packages/unenv-preset/src/index.ts:2]: I added your 2 comments to the tracking issue #7366

The idea is to get this in to unblock the next steps and then we can tweak.

File: packages/unenv-preset/src/runtime/node/async_hooks/$cloudflare.ts
Comment (vicb) [reply packages/unenv-preset/src/runtime/node/async_hooks/$cloudflare.ts:1]: I added your 2 comments to the tracking issue https://github.com/cloudflare/workers-sdk/issues/7366

File: packages/unenv-preset/src/index.ts
Comment (pi0) [reply packages/unenv-preset/src/index.ts:2]: The issue is that, if we start by having peer dependency of unenv, even if unused, it keeps nitro blocked as it forces a pinned version of unenv. 

While it is actually possible that we keep unenv as a `wrangler` dependency so that all consumers are free to choose their versioning policy for `unenv` (the composer) + their preset compatibility.

Comment (pi0) [reply packages/unenv-preset/src/index.ts:2]: Having peer dep as a range in the future is workable. But we hack a dead-lock situation here.

- unenv v2 won't be released unless we can test it fully against nitro
- unenv-preset won't relax peer dep of unenv until unenv v2 is released
- nitro (nightly) cannot test unenv-preset without setting the pinned dep

I might be missing some of the context in your integration plan and steps but it looks to me, `0.0.1` can be safely released without reexports and peer dependency and tested/integrated against wrangler by composing in wrangler layer and this way starting `0.0.1` we can also start testing it against nitro to finalize both sides.

File: .gitignore
Comment (anonrig) [reply .gitignore:201]: Which tracking issue?

File: packages/unenv-preset/build.config.ts
Comment (petebacondarwin) [inline packages/unenv-preset/build.config.ts:10]: Can we output to `dist/runtime` and then fix up the package.json exports, which would also allow us to remove the `.gitignore` change.

File: packages/unenv-preset/src/index.ts
Comment (petebacondarwin) [inline packages/unenv-preset/src/index.ts:2]: Why do we re-export this?
Do presets normally also export `env()` etc?

**Review by petebacondarwin — COMMENTED** [review]
Will need to update `findPackageNames()` in `validate-changesets.test.ts`.

File: packages/unenv-preset/build.config.ts
Comment (vicb) [reply packages/unenv-preset/build.config.ts:10]: Yes, I can change to output to 2 different dist subfolders

File: packages/unenv-preset/turbo.json
Comment (petebacondarwin) [inline packages/unenv-preset/turbo.json:6]: Now that runtime is output to dist is it enough to just do this?

```suggestion
			"outputs": ["dist/**"]
```

**Review by petebacondarwin — APPROVED** [review]

**Comment by petebacondarwin** [comment]
Still need to fix this test:

> Will need to update findPackageNames() in validate-changesets.test.ts.

File: packages/unenv-preset/turbo.json
Comment (vicb) [reply packages/unenv-preset/turbo.json:6]: good catch!

**Comment by petebacondarwin** [comment]
And appears to be a problem with the lockfile?

File: packages/unenv-preset/package.json
Comment (petebacondarwin) [inline packages/unenv-preset/package.json:60]: If you add this here:

```suggestion
	},
	"workers-sdk": {
		"prerelease": true
	}
```

Then we should auto-generate pre-release builds per PR.

**Comment by petebacondarwin** [comment]
I think to do the first deployment we need to mark it as public access:

```sh
npm publish --access public
```

File: packages/unenv-preset/turbo.json
Comment (petebacondarwin) [reply packages/unenv-preset/turbo.json:6]: ...

**Review by petebacondarwin — APPROVED** [review]
Just this one small nit.

File: packages/unenv-preset/turbo.json
Comment (vicb) [reply packages/unenv-preset/turbo.json]: really fixed this time :)
Thanks!

**Pull request merged**
